gothemis/cell/seal_pw.go

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) 2020 Cossack Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cell
+
+/*
+ #cgo LDFLAGS: -lthemis
+ #include <themis/themis.h>
+*/
+import "C"
+
+import (
+	"github.com/cossacklabs/themis/gothemis/errors"
+)
+
+// SecureCellSealPassphrase is Secure Cell in Seal mode.
+// This is the most secure and easy way to protect stored data.
+//
+// The data is protected by a secret passphrase.
+// You can safely use relatively short, human-readable passwords with this mode.
+//
+// If you do not need to keep the secret in the head and would rather write it into a file,
+// consider using symmetric keys.
+// See ``cell.SealWithKey()'' and ``SecureCellSeal''.
+//
+// Read more about Seal mode here:
+//
+// https://docs.cossacklabs.com/pages/secure-cell-cryptosystem/#seal-mode
+type SecureCellSealPassphrase struct {
+	passphrase string
```

Comment:
Oh, now I see, that it looks inefficient and we should provide a way to convert passphrase to key to allow cache this computation and not calculate key every time...

Reply:
Yeah. Here (and in other wrappers) we keep the original passphrase and pass it to C functions that actually perform KDF on *every individual* encryption or decryption.

It is straightforward, API-wise, to prederive key for encryption and cache it. However, it is not obvious how to handle decryption path which requires not only the passphrase but KDF parameters and salt as well, which both can be different for each message.

Our hope is that users would rarely need passphrase API to encrypt or decrypt massive amounts of data so some inefficiency is okay. If you are only going to use this API to encrypt a master key and then decrypt it relatively rarely then it's okay to pay KDF costs (~200 ms) every time you use this API.

Reply:
imho we can perform kdf and correctly pass context to encryption function in our wrappers,  cache generated key and re-use it. Add semi-private (exportable as C code, but not documented anywhere and used only by our wrappers instead users) functions. One, which perform kdf on passphrase and export kdf context + new key. And second, is function for encryption which accept kdf context and symmetric key, and put context to correct place. Then we can use it in our wrappers and store in object symmetric key instead passphrase.

Reply:
Yeah, this idea has been explored (RFC 2, “Detached key derivation”). However, it introduces a lot of open questions:

- API, ABI, data formats need to be backwards compatible
- decryption path seems to be optimizable only with user assistance
- making KDF parameters visible is not “simple” for users
- reusing KDF parameters for multiple messages may have security implications

Immediate benefits include improved performance when encrypting data with the same KDF parameters and a later option to configure KDF for the users.

Given our expectations, this will not improve main user story for passphrase API which we expect to be used relatively rare to decrypt some master secret or a single configuration. It is expected that passphrase API will not be used to decrypt a lot of data chunks (e.g., an passphrased AcraStructs for each individual database cell). With these assumptions caching KDF results is not going to win much performance.

We can introduce optimizations later while keeping high-level API unchanged. The idea is to flesh out passphrase API, get some user input, and then maybe optimize Themis Core requests we do to implement that API.